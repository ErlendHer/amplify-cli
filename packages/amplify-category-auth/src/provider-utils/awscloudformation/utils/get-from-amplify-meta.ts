import { stateManager } from '@aws-amplify/amplify-cli-core';

export const getUserPoolIdFromMeta = (resourceName: string): string | undefined => {
  const amplifyMeta = stateManager.getMeta();
  const resource = (amplifyMeta?.auth || {})[resourceName];

  return resource?.output?.UserPoolId;
};
