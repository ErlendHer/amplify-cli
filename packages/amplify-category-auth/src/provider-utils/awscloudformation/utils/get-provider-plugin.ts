import { $TSContext } from '@aws-amplify/amplify-cli-core';
import { getSupportedServices } from '../../supported-services';
import { ProviderUtils } from '../import/types';

export const getProviderPlugin = (context: $TSContext): ProviderUtils => {
  const serviceMetaData = getSupportedServices().Cognito;
  const { provider } = serviceMetaData;

  return context.amplify.getPluginInstance(context, provider);
};
