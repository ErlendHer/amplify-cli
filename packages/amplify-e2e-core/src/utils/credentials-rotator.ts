import { AssumeRoleCommand, STSClient } from '@aws-sdk/client-sts';
import { fromContainerMetadata } from '@aws-sdk/credential-providers';
import { generateRandomShortId, TEST_PROFILE_NAME } from './index';
import * as ini from 'ini';
import * as fs from 'fs-extra';
import { pathManager } from '@aws-amplify/amplify-cli-core';

const refreshCredentials = async () => {
  try {
    const client = new STSClient({
      // Use CodeBuild role to assume test account role. I.e. don't read credentials from process.env
      credentials: fromContainerMetadata(),
    });
    const roleArn = process.env.TEST_ACCOUNT_ROLE;
    const sessionName = `testSession${generateRandomShortId()}`;
    const command = new AssumeRoleCommand({
      RoleArn: roleArn,
      RoleSessionName: sessionName,
      DurationSeconds: 3600,
    });
    const response = await client.send(command);

    const profileName = TEST_PROFILE_NAME;
    const credentialsContents = ini.parse((await fs.readFile(pathManager.getAWSCredentialsFilePath())).toString());
    credentialsContents[profileName] = credentialsContents[profileName] || {};
    credentialsContents[profileName].aws_access_key_id = response.Credentials.AccessKeyId;
    credentialsContents[profileName].aws_secret_access_key = response.Credentials.SecretAccessKey;
    credentialsContents[profileName].aws_session_token = response.Credentials.SessionToken;
    process.env.AWS_ACCESS_KEY_ID = response.Credentials.AccessKeyId;
    process.env.AWS_SECRET_ACCESS_KEY = response.Credentials.SecretAccessKey;
    process.env.AWS_SESSION_TOKEN = response.Credentials.SessionToken;
    await fs.writeFile(pathManager.getAWSCredentialsFilePath(), ini.stringify(credentialsContents));
    console.log('Test profile credentials refreshed');
  } catch (e) {
    console.error('Test profile credentials request failed');
    console.error(e);
  }
};

export const tryScheduleCredentialRefresh = () => {
  // Early return if CI environment is not Codebuild
  if (!process.env.CI || !process.env.CODEBUILD) {
    return false;
  }

  setInterval(() => {
    void refreshCredentials();
  }, 10 * 60 * 1000);

  return true;
};
