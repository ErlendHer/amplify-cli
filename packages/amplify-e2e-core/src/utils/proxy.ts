import http, { RequestListener, Server } from 'http';
import httpProxy from 'http-proxy';

export class TestProxy {
  proxy;
  server: Server;
  constructor(readonly port = 80) {
    process.env.HTTP_PROXY = `http://localhost:${this.port}`;
    this.proxy = httpProxy.createProxyServer({ target: process.env.HTTP_PROXY });

    this.proxy.on('error', (e) => {
      console.error(e?.message || e);
    });

    console.log(`listening on port ${this.port}`);
    this.server = http.createServer(this.onRequest).listen(this.port);
  }

  onRequest: RequestListener = (req, res) => {
    console.log('serve: ' + req.url);
    // res.writeHead(200, { 'Content-Type': 'text/plain' });
    // res.write('request successfully proxied!' + '\n' + JSON.stringify(req.headers, null, 2));
    // res.end();
    this.proxy.web(req, res, { target: process.env.HTTP_PROXY });
  };

  stop() {
    this.server.close();
    this.proxy.close();
    delete process.env.HTTP_PROXY;
  }
}
