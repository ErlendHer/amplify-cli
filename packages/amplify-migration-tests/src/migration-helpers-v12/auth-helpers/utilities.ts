import type { $TSObject } from '@aws-amplify/amplify-cli-core';
import {
  addAuthUserPoolOnlyWithOAuth,
  AddAuthUserPoolOnlyWithOAuthSettings,
  amplifyPushAuth,
  generateRandomShortId,
} from '@aws-amplify/amplify-e2e-core';

/**
 * sets up a project with auth (UserPool only or UserPool & IdentityPool)
 */
export const setupOgProjectWithAuth = async (ogProjectRoot: string, ogProjectSettings: { name: string }): Promise<void> => {
  const ogShortId = generateRandomShortId();
  await addAuthUserPoolOnlyWithOAuth(ogProjectRoot, createUserPoolWithOAuthSettings(ogProjectSettings.name, ogShortId));
  await amplifyPushAuth(ogProjectRoot);
};

const createUserPoolWithOAuthSettings = (projectPrefix: string, shortId: string): AddAuthUserPoolOnlyWithOAuthSettings => {
  return {
    // eslint-disable spellcheck/spell-checker
    resourceName: `${projectPrefix}oares${shortId}`,
    userPoolName: `${projectPrefix}oaup${shortId}`,
    domainPrefix: `${projectPrefix}oadom${shortId}`,
    signInUrl1: 'https://sin1/',
    signInUrl2: 'https://sin2/',
    signOutUrl1: 'https://sout1/',
    signOutUrl2: 'https://sout2/',
    facebookAppId: 'facebookAppId',
    facebookAppSecret: 'facebookAppSecret',
    googleAppId: 'googleAppId',
    googleAppSecret: 'googleAppSecret',
    amazonAppId: 'amazonAppId',
    amazonAppSecret: 'amazonAppSecret',
    appleAppClientId: 'com.fake.app',
    appleAppTeamId: '2QLEWNDK6K',
    appleAppKeyID: '2QLZXKYJ8J',
    appleAppPrivateKey:
      '----BEGIN PRIVATE KEY----MIGTAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBHkwdwIBAQQgIltgNsTgTfSzUadYiCS0VYtDDMFln/J8i1yJsSIw5g+gCgYIKoZIzj0DAQehRANCAASI8E0L/DhR/mIfTT07v3VwQu6q8I76lgn7kFhT0HvWoLuHKGQFcFkXXCgztgBrprzd419mUChAnKE6y89bWcNw----END PRIVATE KEY----',
    // eslint-enable spellcheck/spell-checker
  };
};

const lambdaCalloutFilter = (r: $TSObject) => r?.Type === 'AWS::Lambda::Function' || r?.Type === 'Custom::LambdaCallout';

const getLambdasInCfnTemplate = (template: $TSObject): $TSObject[] => {
  const lambdaResources = Object.values(template.Resources).filter(lambdaCalloutFilter);
  return lambdaResources;
};

export const expectLambdasInCfnTemplate = (template: $TSObject): void => {
  expect(template?.Resources).toBeDefined();
  const lambdaResources = getLambdasInCfnTemplate(template);
  expect(lambdaResources.length).not.toBe(0);
};

export const expectNoLambdasInCfnTemplate = (template: $TSObject): void => {
  expect(template?.Resources).toBeDefined();
  expect(Object.values(template.Resources).filter(lambdaCalloutFilter).length).toBe(0);
};
