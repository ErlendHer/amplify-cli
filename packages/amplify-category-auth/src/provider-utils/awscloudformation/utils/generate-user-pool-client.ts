import { $TSContext } from '@aws-amplify/amplify-cli-core';
import { ICognitoUserPoolService } from '@aws-amplify/amplify-util-import';
import { getProviderPlugin } from './get-provider-plugin';

export const generateUserPoolClient = async (context: $TSContext): Promise<ICognitoUserPoolService> => {
  const providerPlugin = getProviderPlugin(context);
  const cognitoUserPoolService = await providerPlugin.createCognitoUserPoolService(context);

  return cognitoUserPoolService;
};
