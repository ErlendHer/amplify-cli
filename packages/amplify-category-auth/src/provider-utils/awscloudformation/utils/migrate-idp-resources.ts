import { $TSAny, $TSContext, $TSObject, JSONUtilities, pathManager, Template } from '@aws-amplify/amplify-cli-core';
import { ICognitoUserPoolService } from '@aws-amplify/amplify-util-import';
import { generateUserPoolClient } from './generate-user-pool-client';
import { getUserPoolIdFromMeta } from './get-from-amplify-meta';

const { readJson } = JSONUtilities;
const { getCurrentCfnTemplatePathFromBuild } = pathManager;

export const migrateResourcesToCfn = (resourceName: string): boolean => {
  const authCfnTemplatePath = getCurrentCfnTemplatePathFromBuild('auth', resourceName);
  const authCfnTemplate: Template | undefined = readJson(authCfnTemplatePath, { throwIfNotExist: false });
  const lambdaCalloutCreatedInCloud = authCfnTemplate?.Resources?.HostedUIProvidersCustomResource?.Type === 'AWS::Lambda::Function';
  const providerCreatedInCloud = hasHostedProviderResources(authCfnTemplate);

  return lambdaCalloutCreatedInCloud && !providerCreatedInCloud;
};

const hasHostedProviderResources = (authCfnTemplate: Template | undefined): boolean => {
  return (
    authCfnTemplate?.Resources?.HostedUIFacebookProviderResource?.Type === 'AWS::Cognito::UserPoolIdentityProvider' ||
    authCfnTemplate?.Resources?.HostedUIGoogleProviderResource?.Type === 'AWS::Cognito::UserPoolIdentityProvider' ||
    authCfnTemplate?.Resources?.HostedUILoginWithAmazonProviderResource?.Type === 'AWS::Cognito::UserPoolIdentityProvider' ||
    authCfnTemplate?.Resources?.HostedUISignInWithAppleProviderResource?.Type === 'AWS::Cognito::UserPoolIdentityProvider'
  );
};

/*
 * This function iterates over the user-inputted hostedUICreds and makes SDK calls
 * to populate non-user-updated creds in the array.
 */

export const getHostedUIProviderCredsFromCloud = async (
  resourceName: string,
  providerMeta: $TSObject[],
  updatedUIProviderCreds: $TSObject[],
  context?: $TSContext,
): Promise<$TSObject[]> => {
  const userPoolId = getUserPoolIdFromMeta(resourceName);
  const providerCredsArr = [];

  for (const provider of providerMeta) {
    let providerCreds = updatedUIProviderCreds?.find(({ ProviderName }) => ProviderName === provider.ProviderName) || {};
    const hasEmptyCreds = Object.keys(providerCreds).length === 0;
    const hasNotBeenUpdated = providerCreds && Object.keys(providerCreds).length === 1 && 'ProviderName' in providerCreds;

    if ((hasNotBeenUpdated || hasEmptyCreds) && userPoolId && context) {
      const client = await generateUserPoolClient(context);
      providerCreds = (await getProviderCreds(userPoolId, provider.ProviderName, client)) || providerCreds;
      providerCreds.ProviderName = provider.ProviderName;
    }

    providerCredsArr.push(providerCreds);
  }

  return providerCredsArr;
};

export const getProviderCreds = async (userPoolId: string, providerName: string, client: ICognitoUserPoolService): Promise<$TSAny> => {
  return (await client.getUserPoolIdentityProviderDetails(userPoolId, providerName))?.IdentityProvider?.ProviderDetails;
};
