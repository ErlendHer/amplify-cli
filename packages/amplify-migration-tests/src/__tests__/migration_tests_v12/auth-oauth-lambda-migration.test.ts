import {
  AddAuthUserPoolOnlyWithOAuthSettings,
  amplifyPushForce,
  createNewProjectDir,
  deleteProject,
  deleteProjectDir,
  getProjectMeta,
  getSocialIdpProvider,
} from '@aws-amplify/amplify-e2e-core';
import { allowedVersionsToMigrateFrom, versionCheck } from '../../migration-helpers';
import { expectCorrectOAuthSettings, setupOgProjectWithAuth } from '../../migration-helpers-v12/auth-helpers/utilities';
import { initJSProjectWithProfileV12 } from '../../migration-helpers-v12/init';
import { pullPushForceWithLatestCodebaseValidateParameterAndCfnDrift } from '../../migration-helpers/utils';

const defaultsSettings = {
  name: 'authTest',
  disableAmplifyAppCreation: false,
};

describe('amplify add auth...', () => {
  let projRoot: string;
  let oAuthSettings: AddAuthUserPoolOnlyWithOAuthSettings;
  const projectName: string = 'oauthlambdaRemove';

  beforeAll(async () => {
    const migrateFromVersion = { v: 'uninitialized' };
    const migrateToVersion = { v: 'uninitialized' };
    await versionCheck(process.cwd(), false, migrateFromVersion);
    await versionCheck(process.cwd(), true, migrateToVersion);
    console.log(`Test migration from: ${migrateFromVersion.v} to ${migrateToVersion.v}`);
    expect(allowedVersionsToMigrateFrom).toContain(migrateFromVersion.v);
  });

  beforeEach(async () => {
    projRoot = await createNewProjectDir(projectName);
    await initJSProjectWithProfileV12(projRoot, defaultsSettings);
    // creates a userPool only with OauthSetting and pushes Auth
    oAuthSettings = await setupOgProjectWithAuth(projRoot, { name: 'ogauimphea' });
  });

  afterEach(async () => {
    await deleteProject(projRoot);
    deleteProjectDir(projRoot);
  });

  it('...should init an Js project and add Oauth settings with userpool', async () => {
    const projRoot2 = await createNewProjectDir(`${projectName}2`);
    // using amplify push force here as changes are only related to build files
    await pullPushForceWithLatestCodebaseValidateParameterAndCfnDrift(projRoot, projRoot2);
  });

  it('...should preserve Oauth settings on force push with new', async () => {
    await amplifyPushForce(projRoot, true);
    await expectCorrectOAuthSettings(projRoot, oAuthSettings);
  });
});
