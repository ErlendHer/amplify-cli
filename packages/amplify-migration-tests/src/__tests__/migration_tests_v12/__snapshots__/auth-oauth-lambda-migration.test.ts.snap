// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amplify add auth... ...should init an Js project and add Oauth settings with userpool 1`] = `
"IAM Statement Changes
┌───┬─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬────────┬────────────────────────────────────┬───────────────────────────┬───────────┐
│   │ Resource                                                                                                                                                                                    │ Effect │ Action                             │ Principal                 │ Condition │
├───┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────┼────────────────────────────────────┼───────────────────────────┼───────────┤
│ - │ \${UserPool.Arn}                                                                                                                                                                             │ Allow  │ cognito-idp:DescribeUserPoolClient │ AWS:\${UserPoolClientRole} │           │
│ - │ \${UserPool.Arn}                                                                                                                                                                             │ Allow  │ cognito-idp:UpdateUserPoolClient   │ AWS:\${UserPoolClientRole} │           │
├───┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────┼────────────────────────────────────┼───────────────────────────┼───────────┤
│ - │ {"Fn::Sub":["arn:aws:logs:\${region}:\${account}:log-group:/aws/lambda/\${lambda}:log-stream:*",{"region":"\${AWS::Region}","account":"\${AWS::AccountId}","lambda":"\${OAuthCustomResource}"}]}  │ Allow  │ logs:CreateLogGroup                │ AWS:\${UserPoolClientRole} │           │
│   │                                                                                                                                                                                             │        │ logs:CreateLogStream               │                           │           │
│   │                                                                                                                                                                                             │        │ logs:PutLogEvents                  │                           │           │
├───┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────┼────────────────────────────────────┼───────────────────────────┼───────────┤
│ - │ {"Fn::Sub":["arn:aws:logs:\${region}:\${account}:log-group:/aws/lambda/\${lambda}:log-stream:*",{"region":"\${AWS::Region}","account":"\${AWS::AccountId}","lambda":"\${UserPoolClientLambda}"}]} │ Allow  │ logs:CreateLogGroup                │ AWS:\${UserPoolClientRole} │           │
│   │                                                                                                                                                                                             │        │ logs:CreateLogStream               │                           │           │
│   │                                                                                                                                                                                             │        │ logs:PutLogEvents                  │                           │           │
└───┴─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴────────┴────────────────────────────────────┴───────────────────────────┴───────────┘
(NOTE: There may be security-related changes not in this list. See https://github.com/aws/aws-cdk/issues/1299)

Conditions
[-] Condition ShouldOutputAppClientSecrets: {"Fn::Equals":[{"Ref":"userpoolClientGenerateSecret"},true]}

Resources
[-] AWS::Lambda::Function UserPoolClientLambda destroy
[-] AWS::IAM::Policy UserPoolClientLambdaPolicy destroy
[-] AWS::IAM::Policy UserPoolClientLogPolicy destroy
[-] Custom::LambdaCallout UserPoolClientInputs destroy
[-] AWS::Lambda::Function OAuthCustomResource destroy
[-] AWS::IAM::Policy OAuthCustomResourcePolicy destroy
[-] AWS::IAM::Policy OAuthCustomResourceLogPolicy destroy
[-] Custom::LambdaCallout OAuthCustomResourceInputs destroy
[~] AWS::Cognito::UserPoolClient UserPoolClientWeb 
 ├─ [+] AllowedOAuthFlows
 │   └─ ["code"]
 ├─ [+] AllowedOAuthFlowsUserPoolClient
 │   └─ true
 ├─ [+] AllowedOAuthScopes
 │   └─ ["phone","email","openid","profile","aws.cognito.signin.user.admin"]
 ├─ [+] CallbackURLs
 │   └─ ["https://sin1/","https://sin2/"]
 ├─ [+] LogoutURLs
 │   └─ ["https://sout1/","https://sout2/"]
 ├─ [+] SupportedIdentityProviders
 │   └─ ["Facebook","Google","LoginWithAmazon","SignInWithApple","COGNITO"]
 └─ [~] DependsOn
     └─ @@ -1,3 +1,4 @@
        [ ] [
        [+]   "HostedUIProvidersCustomResourceInputs",
        [ ]   "UserPool"
        [ ] ]
[~] AWS::Cognito::UserPoolClient UserPoolClient 
 ├─ [+] AllowedOAuthFlows
 │   └─ ["code"]
 ├─ [+] AllowedOAuthFlowsUserPoolClient
 │   └─ true
 ├─ [+] AllowedOAuthScopes
 │   └─ ["phone","email","openid","profile","aws.cognito.signin.user.admin"]
 ├─ [+] CallbackURLs
 │   └─ ["https://sin1/","https://sin2/"]
 ├─ [+] LogoutURLs
 │   └─ ["https://sout1/","https://sout2/"]
 ├─ [+] SupportedIdentityProviders
 │   └─ ["Facebook","Google","LoginWithAmazon","SignInWithApple","COGNITO"]
 └─ [~] DependsOn
     └─ @@ -1,3 +1,4 @@
        [ ] [
        [+]   "HostedUIProvidersCustomResourceInputs",
        [ ]   "UserPool"
        [ ] ]
[~] AWS::IAM::Role UserPoolClientRole 
 └─ [-] DependsOn
     └─ ["UserPoolClient"]

Outputs
[-] Output AppClientSecret: {"Value":{"Fn::GetAtt":["UserPoolClientInputs","appSecret"]},"Condition":"ShouldOutputAppClientSecrets"}

"
`;

exports[`amplify add auth... ...should init an Js project and add Oauth settings with userpool 2`] = `
"IAM Statement Changes
┌───┬─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┬────────┬────────────────────────────────────┬───────────────────────────┬───────────┐
│   │ Resource                                                                                                                                                                                    │ Effect │ Action                             │ Principal                 │ Condition │
├───┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────┼────────────────────────────────────┼───────────────────────────┼───────────┤
│ - │ \${UserPool.Arn}                                                                                                                                                                             │ Allow  │ cognito-idp:DescribeUserPoolClient │ AWS:\${UserPoolClientRole} │           │
│ - │ \${UserPool.Arn}                                                                                                                                                                             │ Allow  │ cognito-idp:UpdateUserPoolClient   │ AWS:\${UserPoolClientRole} │           │
├───┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────┼────────────────────────────────────┼───────────────────────────┼───────────┤
│ - │ {"Fn::Sub":["arn:aws:logs:\${region}:\${account}:log-group:/aws/lambda/\${lambda}:log-stream:*",{"region":"\${AWS::Region}","account":"\${AWS::AccountId}","lambda":"\${OAuthCustomResource}"}]}  │ Allow  │ logs:CreateLogGroup                │ AWS:\${UserPoolClientRole} │           │
│   │                                                                                                                                                                                             │        │ logs:CreateLogStream               │                           │           │
│   │                                                                                                                                                                                             │        │ logs:PutLogEvents                  │                           │           │
├───┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┼────────┼────────────────────────────────────┼───────────────────────────┼───────────┤
│ - │ {"Fn::Sub":["arn:aws:logs:\${region}:\${account}:log-group:/aws/lambda/\${lambda}:log-stream:*",{"region":"\${AWS::Region}","account":"\${AWS::AccountId}","lambda":"\${UserPoolClientLambda}"}]} │ Allow  │ logs:CreateLogGroup                │ AWS:\${UserPoolClientRole} │           │
│   │                                                                                                                                                                                             │        │ logs:CreateLogStream               │                           │           │
│   │                                                                                                                                                                                             │        │ logs:PutLogEvents                  │                           │           │
└───┴─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┴────────┴────────────────────────────────────┴───────────────────────────┴───────────┘
(NOTE: There may be security-related changes not in this list. See https://github.com/aws/aws-cdk/issues/1299)

Conditions
[-] Condition ShouldOutputAppClientSecrets: {"Fn::Equals":[{"Ref":"userpoolClientGenerateSecret"},true]}

Resources
[-] AWS::Lambda::Function UserPoolClientLambda destroy
[-] AWS::IAM::Policy UserPoolClientLambdaPolicy destroy
[-] AWS::IAM::Policy UserPoolClientLogPolicy destroy
[-] Custom::LambdaCallout UserPoolClientInputs destroy
[-] AWS::Lambda::Function OAuthCustomResource destroy
[-] AWS::IAM::Policy OAuthCustomResourcePolicy destroy
[-] AWS::IAM::Policy OAuthCustomResourceLogPolicy destroy
[-] Custom::LambdaCallout OAuthCustomResourceInputs destroy
[~] AWS::Cognito::UserPoolClient UserPoolClientWeb 
 ├─ [+] AllowedOAuthFlows
 │   └─ ["code"]
 ├─ [+] AllowedOAuthFlowsUserPoolClient
 │   └─ true
 ├─ [+] AllowedOAuthScopes
 │   └─ ["phone","email","openid","profile","aws.cognito.signin.user.admin"]
 ├─ [+] CallbackURLs
 │   └─ ["https://sin1/","https://sin2/"]
 ├─ [+] LogoutURLs
 │   └─ ["https://sout1/","https://sout2/"]
 ├─ [+] SupportedIdentityProviders
 │   └─ ["Facebook","Google","LoginWithAmazon","SignInWithApple","COGNITO"]
 └─ [~] DependsOn
     └─ @@ -1,3 +1,4 @@
        [ ] [
        [+]   "HostedUIProvidersCustomResourceInputs",
        [ ]   "UserPool"
        [ ] ]
[~] AWS::Cognito::UserPoolClient UserPoolClient 
 ├─ [+] AllowedOAuthFlows
 │   └─ ["code"]
 ├─ [+] AllowedOAuthFlowsUserPoolClient
 │   └─ true
 ├─ [+] AllowedOAuthScopes
 │   └─ ["phone","email","openid","profile","aws.cognito.signin.user.admin"]
 ├─ [+] CallbackURLs
 │   └─ ["https://sin1/","https://sin2/"]
 ├─ [+] LogoutURLs
 │   └─ ["https://sout1/","https://sout2/"]
 ├─ [+] SupportedIdentityProviders
 │   └─ ["Facebook","Google","LoginWithAmazon","SignInWithApple","COGNITO"]
 └─ [~] DependsOn
     └─ @@ -1,3 +1,4 @@
        [ ] [
        [+]   "HostedUIProvidersCustomResourceInputs",
        [ ]   "UserPool"
        [ ] ]
[~] AWS::IAM::Role UserPoolClientRole 
 └─ [-] DependsOn
     └─ ["UserPoolClient"]

Outputs
[-] Output AppClientSecret: {"Value":{"Fn::GetAtt":["UserPoolClientInputs","appSecret"]},"Condition":"ShouldOutputAppClientSecrets"}

"
`;
